File: workflow/kernel/amanar_workflow/receipts.py
"""Source snapshots and receipt freshness checks."""
from __future__ import annotations

import hashlib
import json
import os
import re
import stat
import subprocess
from datetime import datetime
from pathlib import Path
from typing import Any

from .contract import check_hash, path_in, workflow_hash
from .errors import EvidenceError, WorkflowError
from .execution import MAX_OUTPUT, parse_tests

RUNTIME_PREFIX = ".amanar/run/"


def _git(root: Path, *args: str) -> bytes:
    result = subprocess.run(
        ["git", *args], cwd=root, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
    )
    if result.returncode != 0:
        message = result.stderr.decode("utf-8", errors="replace").strip()
        raise WorkflowError(f"Git source inspection failed: {message}")
    return result.stdout


def _file_digest(path: Path) -> str:
    digest = hashlib.sha256()
    if path.is_symlink():
        digest.update(b"symlink\0")
        digest.update(os.readlink(path).encode("utf-8", errors="surrogateescape"))
        return digest.hexdigest()
    with path.open("rb") as stream:
        for chunk in iter(lambda: stream.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def _entry_digest(path: Path) -> str:
    details = path.lstat()
    if stat.S_ISDIR(details.st_mode):
        return "directory"
    if stat.S_ISREG(details.st_mode) or stat.S_ISLNK(details.st_mode):
        return _file_digest(path)
    return f"special:{stat.S_IFMT(details.st_mode):o}"


def _filesystem_entries(root: Path) -> dict[str, str]:
    files: dict[str, str] = {}

    def visit(directory: Path, prefix: str = "") -> None:
        try:
            entries = sorted(os.scandir(directory), key=lambda item: item.name)
        except OSError as exc:
            raise WorkflowError(f"filesystem source inspection failed: {exc}") from exc
        for entry in entries:
            name = f"{prefix}/{entry.name}" if prefix else entry.name
            if name == ".git" or name.startswith(".git/"):
                continue
            if name == ".amanar/run" or name.startswith(RUNTIME_PREFIX):
                continue
            path = Path(entry.path)
            if entry.is_dir(follow_symlinks=False):
                visit(path, name)
            else:
                files[name] = _entry_digest(path)

    visit(root)
    return files


def source_snapshot(root: Path) -> dict[str, Any]:
    head = _git(root, "rev-parse", "HEAD").decode().strip()
    names = _git(root, "ls-files", "-z", "--cached")
    files = _filesystem_entries(root)
    for raw in names.split(b"\0"):
        if not raw:
            continue
        name = raw.decode("utf-8", errors="surrogateescape")
        if name == ".amanar/run" or name.startswith(RUNTIME_PREFIX):
            continue
        if name not in files:
            files[name] = "missing"
    payload = {"head": head, "files": files}
    encoded = json.dumps(payload, sort_keys=True, separators=(",", ":")).encode()
    payload["digest"] = hashlib.sha256(encoded).hexdigest()
    return payload


def changed_paths(baseline: dict[str, Any], current: dict[str, Any]) -> list[str]:
    before = baseline["files"]
    after = current["files"]
    return sorted(path for path in set(before) | set(after) if before.get(path) != after.get(path))


def assert_scope(contract: dict[str, Any], baseline: dict[str, Any], current: dict[str, Any]) -> None:
    if baseline["head"] != current["head"]:
        raise EvidenceError("source HEAD changed since begin")
    for path in changed_paths(baseline, current):
        if any(path_in(path, item) for item in contract["exclusions"]):
            raise EvidenceError(f"excluded path changed: {path}")
        if not any(path_in(path, item) for item in contract["scope"]):
            raise EvidenceError(f"out-of-scope path changed: {path}")


def assert_artifacts(root: Path, contract: dict[str, Any]) -> None:
    missing = [path for path in contract["artifacts"] if not (root / path.rstrip("/")).exists()]
    if missing:
        raise EvidenceError(f"declared artifacts missing: {', '.join(missing)}")


def receipt_problem(
    receipt: dict[str, Any], contract: dict[str, Any], check: dict[str, Any], source_digest: str,
) -> str | None:
    required = {
        "receiptVersion", "workflowId", "workflowHash", "checkId", "checkDefinitionHash",
        "sourceDigest", "command", "exitCode", "discoveredTests", "stdoutSha256",
        "stderrSha256", "stdoutTruncated", "stderrTruncated", "timedOut", "passed",
        "recordedAt",
    }
    if set(receipt) != required:
        return f"{check['id']} receipt fields are invalid"
    if receipt.get("receiptVersion") != "1.0.0" or receipt.get("workflowId") != contract["id"]:
        return f"{check['id']} receipt identity is invalid"
    if any(not re.fullmatch(r"[0-9a-f]{64}", str(receipt.get(field, "")))
           for field in ("workflowHash", "checkDefinitionHash", "sourceDigest", "stdoutSha256", "stderrSha256")):
        return f"{check['id']} receipt digest is invalid"
    if type(receipt.get("discoveredTests")) not in {int, type(None)}:
        return f"{check['id']} receipt test count is invalid"
    if any(type(receipt.get(field)) is not bool for field in
           ("stdoutTruncated", "stderrTruncated", "timedOut", "passed")):
        return f"{check['id']} receipt boolean is invalid"
    try:
        datetime.fromisoformat(receipt["recordedAt"])
    except (TypeError, ValueError):
        return f"{check['id']} receipt timestamp is invalid"
    expected = {
        "workflowHash": workflow_hash(contract),
        "checkDefinitionHash": check_hash(check),
        "checkId": check["id"],
        "command": check["command"],
        "sourceDigest": source_digest,
    }
    for key, value in expected.items():
        if receipt.get(key) != value:
            return f"{check['id']} receipt has stale {key}"
    if receipt.get("passed") is not True:
        return f"{check['id']} receipt did not pass"
    if receipt.get("exitCode") != check["expectedExit"]:
        return f"{check['id']} receipt exit code is stale"
    discovered = receipt.get("discoveredTests")
    if discovered is None or discovered < check["minTests"]:
        return f"{check['id']} receipt test count is insufficient"
    return None


def output_problem(receipt: dict[str, Any], check: dict[str, Any], output_dir: Path) -> str | None:
    streams: dict[str, bytes] = {}
    for stream in ("stdout", "stderr"):
        path = output_dir / f"{check['id']}.{stream}"
        try:
            details = path.lstat()
            if not stat.S_ISREG(details.st_mode) or details.st_size > MAX_OUTPUT:
                raise OSError("not a bounded regular file")
            streams[stream] = path.read_bytes()
        except OSError:
            return f"{check['id']} stored output is missing or invalid"
        actual = hashlib.sha256(streams[stream]).hexdigest()
        if receipt.get(f"{stream}Sha256") != actual:
            return f"{check['id']} {stream} digest does not match stored output"
    combined = streams["stdout"].decode("utf-8", errors="replace") + "\n" + streams["stderr"].decode(
        "utf-8", errors="replace",
    )
    discovered = parse_tests(check["testParser"], combined)
    if discovered != receipt.get("discoveredTests"):
        return f"{check['id']} stored output test count does not match receipt"
    if not all(token in combined for token in check["outputContains"]):
        return f"{check['id']} stored output lacks required tokens"
    passed = (
        receipt.get("timedOut") is False
        and receipt.get("exitCode") == check["expectedExit"]
        and discovered is not None
        and discovered >= check["minTests"]
    )
    if receipt.get("passed") is not passed:
        return f"{check['id']} receipt outcome does not match stored output"
    return None

File: src/kernel/src/receipts.ts
/**
 * Source snapshots and receipt freshness checks.
 * Mirrors amanar_workflow/receipts.py exactly.
 */

import crypto from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';
import { spawnSync } from 'node:child_process';
import { checkHash, pathIn, workflowHash } from './contract.ts';
import { EvidenceError, WorkflowError } from './errors.ts';
import { MAX_OUTPUT, parseTests } from './execution.ts';

export const RUNTIME_PREFIX = '.amanar/run/';

function git(root: string, ...args: string[]): Buffer {
  const result = spawnSync('git', args, {
    cwd: root,
    stdio: ['ignore', 'pipe', 'pipe'],
  });
  if (result.status !== 0) {
    const message = (result.stderr as Buffer).toString('utf8', 0).trim();
    throw new WorkflowError(`Git source inspection failed: ${message}`);
  }
  return result.stdout as Buffer;
}

function fileDigest(filePath: string): string {
  const lstat = fs.lstatSync(filePath);
  if (lstat.isSymbolicLink()) {
    const h = crypto.createHash('sha256');
    h.update('symlink\0', 'utf8');
    // readlink with buffer gives raw bytes, matching Python's os.readlink(...).encode(errors="surrogateescape")
    const linkTarget = fs.readlinkSync(filePath, { encoding: 'buffer' });
    h.update(linkTarget);
    return h.digest('hex');
  }
  const data = fs.readFileSync(filePath);
  return crypto.createHash('sha256').update(data).digest('hex');
}

export function _fileDigest(filePath: string): string {
  return fileDigest(filePath);
}

function entryDigest(filePath: string): string {
  const stat = fs.lstatSync(filePath);
  if (stat.isDirectory()) return 'directory';
  if (stat.isFile() || stat.isSymbolicLink()) return fileDigest(filePath);
  const ifmt = stat.mode & 0o170000;
  return `special:${ifmt.toString(8)}`;
}

function filesystemEntries(root: string): Record<string, string> {
  const files: Record<string, string> = {};

  function visit(directory: string, prefix: string): void {
    let entries: fs.Dirent[];
    try {
      entries = fs.readdirSync(directory, { withFileTypes: true });
    } catch (e) {
      throw new WorkflowError(`filesystem source inspection failed: ${e}`);
    }
    entries.sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));

    for (const entry of entries) {
      const name = prefix ? `${prefix}/${entry.name}` : entry.name;
      if (name === '.git' || name.startsWith('.git/')) continue;
      if (name === '.amanar/run' || name.startsWith(RUNTIME_PREFIX)) continue;

      const entryPath = path.join(directory, entry.name);
      // isDirectory() on Dirent does NOT follow symlinks
      if (entry.isDirectory()) {
        visit(entryPath, name);
      } else {
        files[name] = entryDigest(entryPath);
      }
    }
  }

  visit(root, '');
  return files;
}

/**
 * Stable JSON with sorted keys (no spaces) — matches Python's default json.dumps
 * used for snapshot encoding (ensure_ascii=True behavior for non-ASCII chars).
 */
function snapshotJson(value: unknown): string {
  function encode(v: unknown): string {
    if (v === null) return 'null';
    if (typeof v === 'boolean') return v ? 'true' : 'false';
    if (typeof v === 'number') return JSON.stringify(v);
    if (typeof v === 'string') {
      // Escape non-ASCII characters like Python's ensure_ascii=True
      return JSON.stringify(v).replace(/[^\x00-\x7F]/gu, c => {
        const cp = c.codePointAt(0)!;
        if (cp <= 0xffff) return `\\u${cp.toString(16).padStart(4, '0')}`;
        // Surrogate pair for code points > 0xffff
        const hi = Math.floor((cp - 0x10000) / 0x400) + 0xd800;
        const lo = ((cp - 0x10000) % 0x400) + 0xdc00;
        return `\\u${hi.toString(16).padStart(4, '0')}\\u${lo.toString(16).padStart(4, '0')}`;
      });
    }
    if (Array.isArray(v)) return '[' + v.map(encode).join(',') + ']';
    if (typeof v === 'object') {
      const obj = v as Record<string, unknown>;
      const keys = Object.keys(obj).sort();
      return '{' + keys.map(k => encode(k) + ':' + encode(obj[k])).join(',') + '}';
    }
    return JSON.stringify(v);
  }
  return encode(value);
}

export function sourceSnapshot(root: string): Record<string, unknown> {
  const head = git(root, 'rev-parse', 'HEAD').toString('utf8').trim();
  const namesRaw = git(root, 'ls-files', '-z', '--cached');
  const files = filesystemEntries(root);

  // namesRaw is null-terminated list of tracked file paths
  const names = namesRaw.toString('utf8').split('\0');
  for (const name of names) {
    if (!name) continue;
    if (name === '.amanar/run' || name.startsWith(RUNTIME_PREFIX)) continue;
    if (!(name in files)) {
      files[name] = 'missing';
    }
  }

  const payload: Record<string, unknown> = { head, files };
  const encoded = Buffer.from(snapshotJson(payload), 'utf8');
  payload['digest'] = crypto.createHash('sha256').update(encoded).digest('hex');
  return payload;
}

export function changedPaths(
  baseline: Record<string, unknown>,
  current: Record<string, unknown>,
): string[] {
  const before = baseline['files'] as Record<string, string>;
  const after = current['files'] as Record<string, string>;
  const allPaths = new Set([...Object.keys(before), ...Object.keys(after)]);
  return [...allPaths].filter(p => before[p] !== after[p]).sort();
}

export function assertScope(
  contract: Record<string, unknown>,
  baseline: Record<string, unknown>,
  current: Record<string, unknown>,
): void {
  if (baseline['head'] !== current['head']) {
    throw new EvidenceError('source HEAD changed since begin');
  }
  for (const filePath of changedPaths(baseline, current)) {
    if ((contract['exclusions'] as string[]).some(item => pathIn(filePath, item))) {
      throw new EvidenceError(`excluded path changed: ${filePath}`);
    }
    if (!(contract['scope'] as string[]).some(item => pathIn(filePath, item))) {
      throw new EvidenceError(`out-of-scope path changed: ${filePath}`);
    }
  }
}

export function assertArtifacts(root: string, contract: Record<string, unknown>): void {
  const missing = (contract['artifacts'] as string[]).filter(
    p => !fs.existsSync(path.join(root, p.replace(/\/$/, ''))),
  );
  if (missing.length > 0) {
    throw new EvidenceError(`declared artifacts missing: ${missing.join(', ')}`);
  }
}

const HEX64 = /^[0-9a-f]{64}$/;

export function receiptProblem(
  receipt: Record<string, unknown>,
  contract: Record<string, unknown>,
  check: Record<string, unknown>,
  sourceDigest: string,
): string | null {
  const required = new Set([
    'receiptVersion', 'workflowId', 'workflowHash', 'checkId', 'checkDefinitionHash',
    'sourceDigest', 'command', 'exitCode', 'discoveredTests', 'stdoutSha256',
    'stderrSha256', 'stdoutTruncated', 'stderrTruncated', 'timedOut', 'passed',
    'recordedAt',
  ]);
  if (
    new Set(Object.keys(receipt)).size !== required.size ||
    !Object.keys(receipt).every(k => required.has(k))
  ) {
    return `${check['id']} receipt fields are invalid`;
  }

  if (receipt['receiptVersion'] !== '1.0.0' || receipt['workflowId'] !== contract['id']) {
    return `${check['id']} receipt identity is invalid`;
  }

  for (const field of ['workflowHash', 'checkDefinitionHash', 'sourceDigest', 'stdoutSha256', 'stderrSha256']) {
    if (!HEX64.test(String(receipt[field] ?? ''))) {
      return `${check['id']} receipt digest is invalid`;
    }
  }

  const discoveredRaw = receipt['discoveredTests'];
  if (typeof discoveredRaw !== 'number' && discoveredRaw !== null) {
    return `${check['id']} receipt test count is invalid`;
  }

  for (const field of ['stdoutTruncated', 'stderrTruncated', 'timedOut', 'passed']) {
    if (typeof receipt[field] !== 'boolean') {
      return `${check['id']} receipt boolean is invalid`;
    }
  }

  try {
    const d = new Date(receipt['recordedAt'] as string);
    if (isNaN(d.getTime())) throw new Error('invalid');
  } catch {
    return `${check['id']} receipt timestamp is invalid`;
  }

  const expected: Record<string, unknown> = {
    workflowHash: workflowHash(contract),
    checkDefinitionHash: checkHash(check),
    checkId: check['id'],
    command: check['command'],
    sourceDigest,
  };
  for (const [key, value] of Object.entries(expected)) {
    if (receipt[key] !== value) {
      return `${check['id']} receipt has stale ${key}`;
    }
  }

  if (receipt['passed'] !== true) {
    return `${check['id']} receipt did not pass`;
  }
  if (receipt['exitCode'] !== check['expectedExit']) {
    return `${check['id']} receipt exit code is stale`;
  }

  const discovered = receipt['discoveredTests'] as number | null;
  if (discovered === null || discovered < (check['minTests'] as number)) {
    return `${check['id']} receipt test count is insufficient`;
  }

  return null;
}

export function outputProblem(
  receipt: Record<string, unknown>,
  check: Record<string, unknown>,
  outputDir: string,
): string | null {
  const streams: Record<string, Buffer> = {};
  for (const stream of ['stdout', 'stderr'] as const) {
    const filePath = path.join(outputDir, `${check['id']}.${stream}`);
    try {
      const stat = fs.lstatSync(filePath);
      if (!stat.isFile() || stat.size > MAX_OUTPUT) {
        throw new Error('not a bounded regular file');
      }
      streams[stream] = fs.readFileSync(filePath);
    } catch {
      return `${check['id']} stored output is missing or invalid`;
    }
    const actual = crypto.createHash('sha256').update(streams[stream]).digest('hex');
    if (receipt[`${stream}Sha256`] !== actual) {
      return `${check['id']} ${stream} digest does not match stored output`;
    }
  }

  const combined = streams['stdout'].toString('utf8') + '\n' + streams['stderr'].toString('utf8');
  const discovered = parseTests(check['testParser'] as string, combined);

  if (discovered !== receipt['discoveredTests']) {
    return `${check['id']} stored output test count does not match receipt`;
  }
  if (!(check['outputContains'] as string[]).every(token => combined.includes(token))) {
    return `${check['id']} stored output lacks required tokens`;
  }

  const passed =
    receipt['timedOut'] === false &&
    receipt['exitCode'] === check['expectedExit'] &&
    discovered !== null &&
    (discovered as number) >= (check['minTests'] as number);

  if (receipt['passed'] !== passed) {
    return `${check['id']} receipt outcome does not match stored output`;
  }

  return null;
}
